oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/base/IoTDBInsertRequest.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.skywalking.oap.server.core.storage.StorageData;
+import org.apache.skywalking.oap.server.core.storage.StorageHashMapBuilder;
+import org.apache.skywalking.oap.server.core.storage.type.StorageDataComplexObject;
+import org.apache.skywalking.oap.server.library.client.request.InsertRequest;
+import org.apache.skywalking.oap.server.library.client.request.UpdateRequest;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBClient;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBTableMetaInfo;
+
+@Getter
+@Setter
+@ToString
+@Slf4j
+public class IoTDBInsertRequest implements InsertRequest, UpdateRequest {
+    private String modelName;
+    private long time;
+    private List<String> indexes;
+    private List<String> indexValues;
+    private List<String> measurements;
+    private List<TSDataType> measurementTypes;
+    private List<Object> measurementValues;
+
+    public <T extends StorageData> IoTDBInsertRequest(String modelName, long time, T storageData,
+                                                      StorageHashMapBuilder<T> storageBuilder) {
+        this.modelName = modelName;
+        this.time = time;
+        indexes = IoTDBTableMetaInfo.get(modelName).getIndexes();
+        indexValues = new ArrayList<>(indexes.size());
+        Map<String, Object> storageMap = storageBuilder.entity2Storage(storageData);
+
+        indexes.forEach(index -> {
+            if (index.equals(IoTDBClient.ID_IDX)) {
+                indexValues.add(storageData.id());
+            } else if (storageMap.containsKey(index)) {
+                indexValues.add(String.valueOf(storageMap.get(index)));
+                storageMap.remove(index);
+            }
+        });
+
+        // time_bucket has changed to time before calling this method, so remove it from measurements
+        storageMap.remove(IoTDBClient.TIME_BUCKET);
+        // processing illegal value
+        Iterator<Map.Entry<String, Object>> entryIterator = storageMap.entrySet().iterator();
+        while (entryIterator.hasNext()) {
+            Map.Entry<String, Object> entry = entryIterator.next();
+            if (entry.getValue() == null) {
+                entryIterator.remove();
+            }
+            if (entry.getValue() instanceof StorageDataComplexObject) {
+                storageMap.put(entry.getKey(), ((StorageDataComplexObject) entry.getValue()).toStorageData());
+            }
+        }
+
+        measurements = new ArrayList<>(storageMap.keySet());
+        Map<String, TSDataType> columnAndTypeMap = IoTDBTableMetaInfo.get(modelName).getColumnAndTypeMap();
+        measurementTypes = new ArrayList<>(measurements.size());
+        for (String measurement : measurements) {
+            measurementTypes.add(columnAndTypeMap.get(measurement));
+        }
+        measurementValues = new ArrayList<>(storageMap.values());
+
+        // processing illegal measurement
```

Comment:
What do you mean illegal? I think this is a wrong word chosen? Maybe you mean to process several special fields which formats unsupported by IoTDB?

Reply:
Ok, I will rewrite it to make it clear.